src/bar.cpp

```diff
@@ -683,6 +710,59 @@ void waybar::Bar::setVisible(bool value) {
 
 void waybar::Bar::toggle() { setVisible(!visible); }
 
+auto waybar::Bar::showMainbar(GtkWidget* widget, const gchar* h, gpointer user_data) -> void {
+  auto curbar = static_cast<Bar*>(user_data);
+  curbar->autohide_mutex.lock();
+
+  // Delete timeout callbacks;
+  auto tag = curbar->autohide_timeout_handle.load();
+  if(tag !=0 && g_source_remove(tag)) {
+    spdlog::debug("removed autohide timeout handle in showMainbar: {}",tag);
+    curbar->autohide_timeout_handle.store(0);
+  }
+
+  spdlog::debug("hotspot: off; bar: on");
+  if(!curbar->window.is_visible()) curbar->window.show_all();
+  if(curbar->hotspotWindow.is_visible()) curbar->hotspotWindow.hide();
+
+  curbar->autohide_mutex.unlock();
+}
+
+auto waybar::Bar::hideMainbarCallback(gpointer user_data) -> void {
+  auto curbar = static_cast<Bar*>(user_data);
+  curbar->autohide_mutex.lock();
+
+  spdlog::debug("hotspot: on; bar: off");
+  if(!curbar->hotspotWindow.is_visible()) curbar->hotspotWindow.show_all();
+  if(curbar->window.is_visible())curbar->window.hide();
+
+  curbar->autohide_timeout_handle.store(0);
+  curbar->autohide_mutex.unlock();
+}
+
+auto waybar::Bar::hideMainbar(GtkWidget* widget, const char* h, gpointer user_data) -> void {
+  auto curbar = static_cast<Bar*>(user_data);
+  curbar->autohide_mutex.lock();
+  auto tag = curbar->autohide_timeout_handle.load();
+  if(tag !=0 && g_source_remove(tag)) {
+    spdlog::debug("removed timeout handle in hideMainbar: {}",tag);
+    curbar->autohide_timeout_handle.store(0);
+  } else {
+    curbar->autohide_timeout_handle.store(g_timeout_add_once(curbar->autohide_delay_ms.load(),GSourceOnceFunc(hideMainbarCallback),curbar));
```

Comment:
This way of setting the timeout is unaware of the object lifetimes and may trigger the callback after the bar instance is destroyed. Please, use glibmm and sigc::mem_fun.

Reply:
@alebastr , I have corrected the codebase. It uses sigc::mem_fun now.